File: UninstallApplication/app/src/main/java/com/example/uninstallapplication/MainActivity.java
package com.example.uninstallapplication;

import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;

import android.content.Intent;
import android.content.pm.ApplicationInfo;
import android.content.pm.PackageManager;
import android.net.Uri;
import android.os.Bundle;
import android.util.Log;
import android.view.View;
import android.view.ViewGroup;
import android.widget.BaseAdapter;
import android.widget.Button;
import android.widget.ImageView;
import android.widget.ListView;
import android.widget.TextView;
import android.widget.Toast;

import java.util.ArrayList;
import java.util.List;

public class MainActivity extends AppCompatActivity {

    ListView listView;
    List<MainData> dataList = new ArrayList<>();

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        listView = findViewById(R.id.list_view);
        PackageManager manager = getPackageManager();
        List<ApplicationInfo> infoList = manager.getInstalledApplications(
                PackageManager.GET_META_DATA
        );
        for(ApplicationInfo info: infoList){
            if(!((info.flags & ApplicationInfo.FLAG_SYSTEM ) != 0)){
                if(!(info.packageName.equals("com.example.uninstallapplication"))){
//                    Log.e("UA_FOR_EACH","In for each loop");
                    MainData data = new MainData();
                    data.setName(info.loadLabel(manager).toString());
                    data.setPackageName(info.packageName);
                    data.setLogo(info.loadIcon(manager));
                    dataList.add(data);
                }
            }
        }

        listView.setAdapter(new BaseAdapter() {
            @Override
            public int getCount() {
                return dataList.size();
            }

            @Override
            public Object getItem(int i) {
                return i;
            }

            @Override
            public long getItemId(int i) {
                return i;
            }

            @Override
            public View getView(int i, View view, ViewGroup viewGroup) {
                View adapterView = getLayoutInflater().inflate(R.layout.list_row_item,null);
                MainData data = dataList.get(i);
                // Initialize and assign variable
                ImageView ivAppLogo =adapterView.findViewById(R.id.iv_app_logo);
                TextView tvAppName = adapterView.findViewById(R.id.tv_app_name);
                Button btUninstall = adapterView.findViewById(R.id.bt_uninstall);
                ivAppLogo.setImageDrawable(data.getLogo());
                tvAppName.setText(data.getName());
                btUninstall.setOnClickListener(new View.OnClickListener() {
                    @Override
                    public void onClick(View view) {
                        MainData mainData = dataList.get(i);
                        Intent intent = new Intent(Intent.ACTION_UNINSTALL_PACKAGE);
                        intent.setData(Uri.parse("package:"+mainData.getPackageName()));
                        intent.putExtra(Intent.EXTRA_RETURN_RESULT,true);
                        startActivityForResult(intent,100);
                    }
                });
                return adapterView;
            }
        });
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, @Nullable Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        if (requestCode == 100 && resultCode == RESULT_OK) {
            Toast.makeText(getApplicationContext(), "Application Uninstalled", Toast.LENGTH_SHORT).show();
            recreate();
        }else{
            Toast.makeText(getApplicationContext(), "Unable to Uninstall", Toast.LENGTH_SHORT).show();
        }
    }
}

File: calculatorUdemy/app/src/main/java/com/example/calculatorudemy/MainActivity.kt
package com.example.calculatorudemy

import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import android.view.View
import android.widget.Button
import android.widget.TextView
import android.widget.Toast
import java.lang.ArithmeticException

class MainActivity : AppCompatActivity() {
    private var inputBtn: TextView? = null
    var lastDigit: Boolean = false
    var lastDecimal: Boolean = false
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        inputBtn = findViewById(R.id.input)
    }
    fun onDigit(view: View){
//        Toast.makeText(this,"clicked",Toast.LENGTH_SHORT).show()
        inputBtn?.append((view as Button).text)
        lastDigit = true
        lastDecimal = false
    }

    fun onClear(view: View){
        inputBtn?.text = ""
    }

    fun onDecimal(view: View){
        if(lastDigit && !lastDecimal){
            inputBtn?.append(".")
            lastDigit = false
            lastDecimal = true
        }
    }

    fun onOperator(view: View){
        inputBtn?.text?.let {
            if(lastDigit && !isOperatorAdded(it.toString())){
                inputBtn?.append((view as Button).text)
                lastDigit = false
                lastDecimal = false
            }
        }
    }

    private fun isOperatorAdded(value: String): Boolean{
        return if(value.startsWith("-")){
            false
        }
        else{
            value.contains("/") || value.contains("*")
                    || value.contains("-") || value.contains("+")
        }
    }

    fun onEqual(view : View){
        if(lastDigit){
            var input = inputBtn?.text.toString()
            var prefix = ""
            try {
                if(input.startsWith("-")){
                    prefix = "-"
                    input = input.substring(1)
                }
                if(input.contains("-")) {
                    var split = input.split("-")
                    var one = split[0]
                    var two = split[1]
                    if(prefix.isNotEmpty()){
                        one = prefix + one
                    }
                    inputBtn?.text = (one.toDouble() - two.toDouble()).toString()
                }else if(input.contains("+")) {
                    var split = input.split("+")
                    var one = split[0]
                    var two = split[1]
                    if(prefix.isNotEmpty()){
                        one = prefix + one
                    }
                    inputBtn?.text = (one.toDouble() + two.toDouble()).toString()
                }else if(input.contains("*")) {
                    var split = input.split("*")
                    var one = split[0]
                    var two = split[1]
                    if(prefix.isNotEmpty()){
                        one = prefix + one
                    }
                    inputBtn?.text = (one.toDouble() * two.toDouble()).toString()
                }else if(input.contains("/")) {
                    var split = input.split("/")
                    var one = split[0]
                    var two = split[1]
                    if(prefix.isNotEmpty()){
                        one = prefix + one
                    }
                    inputBtn?.text = (one.toDouble() / two.toDouble()).toString()
                }
            }catch (e: ArithmeticException){
                e.printStackTrace()
            }

        }
    }
}